### ISSUE TITLE
Gymleaders

### ISSUE THREAD

**Issue by WorldwideErrors** [PokeAPI/pokeapi#580]
Hi, maybe it's an idea to add the gymleaders of the different generations. You can then fetch their names, types, team and badges. This would be really helpful!

**Comment by jmorag** [comment]
I might be interested in taking this on. Would an interface like this under `/gym-leader/{id or name}` be reasonable?
```typescript
interface GymLeader {
  id: number;
  name: string;
  version: { name; url };
  team: {
    pokemon: { name; url };
    level: number;
    move: { name; url }[];
    ability: { name; url };
    held_item: { name; url };
  }[];
  reward: number;
}
```
The data can be found at pages like https://bulbapedia.bulbagarden.net/wiki/Wallace#Pok.C3.A9mon_Emerald.

**Comment by magical** [comment]
If this is added it should be for all trainers, not just gym leaders. Also you'll need to account for rematches.

**Comment by jmorag** [comment]
Do you mean all trainers in the entire game or just the trainers in the gym before the leader? In either case, the interface would have to be slightly different, as I don't know where to get movesets for the pokemon belonging to non-leader trainers. Maybe it would make sense to do gym leaders and normal trainers separately since normal trainers don't change pokemon on rematch (that I know of).

**Comment by magical** [comment]
> Do you mean all trainers in the entire game or just the trainers in the gym before the leader?

The former. Gym leaders are just a special type of trainer (they're literally stored the same way in the game). Whether or not to add the data all at once or in piecemeal is a separate question, but I think the API should be designed with both in mind.

> normal trainers don't change pokemon on rematch (that I know of).

There were plenty of trainers in gen 3 (at least) that would rematch you with a better team. https://bulbapedia.bulbagarden.net/wiki/Hoenn_Route_117#Trainers

**Comment by jmorag** [comment]
You're right. I've found the ROM data for trainers (emerald) [here](https://github.com/pret/pokeemerald/blob/master/src/data/trainer_parties.h). For non-custom learnsets, the code [here](https://github.com/pret/pokeemerald/blob/master/src/pokemon.c#L2971) suggests that pokemon always know the four most recent moves they could have learned. Am I reading that correctly?

**Comment by phalt** [comment]
This would be an awesome data set to integrate but definitely a massive undertaking!

**Comment by laurenarnett** [comment]
I was able to find data for trainers' pokemon movesets on Serebii at links like [here](https://www.serebii.net/pokearth/galar/galarmineno.2.shtml) or [here](https://www.serebii.net/pokearth/unova/route9.shtml). For Emerald on Serebii though, the [movesets are left blank](https://www.serebii.net/pokearth/hoenn/route113.shtml). It seems like data from older gens may not be able to be scraped from Serebii. That data may need to be scraped from the ROM, like @jmorag suggested.

**Comment by BVSTSirop** [comment]
has this ever been implemented?

**Comment by Naramsim** [comment]
No yet

**Comment by programgames** [comment]
I checked serebii route pages and it seems like a lot a trainers data are available, maybe all.
Bur the reward is not available on serebii. Bulbapedia have the rewards data.
The moveset of the traînées is not always available also.
We had to choose which data is more important to serve here.

**Comment by Naramsim** [comment]
If you are willing you could start by defining the API structure, some suggestions are already proposed above.
